### ISSUE TITLE
diagnose: specify the failure-clustering procedure and move deterministic trace parsing into scripts/

### ISSUE THREAD

**Issue by OsherElhadad** [skillberry-ai/cap-evolve#342]
## Summary

`skills/phases/diagnose/` is the most load-bearing skill body in the repo — `harness.py:1231-1243` copies it into `workdir/guidance/diagnose/` and `templates/project/optimizer/INSTRUCTIONS.md:85` tells the optimizer to read it every single iteration — and it is also the least disciplined. The skill asks the model to cluster failures "by shared root cause" (SKILL.md:33, 77-83) without ever defining how a signature is derived or when two failures belong to the same cluster; the only concrete procedure in the package, `normalized_feedback_signature` (`scripts/run.py:49-60`), is never mentioned in SKILL.md and is *unreachable* in production because that same copy step passes `ignore=shutil.ignore_patterns("__pycache__", "scripts", "*.pyc")` (`harness.py:1240`) — so the deterministic clusterer is excluded from the one directory that reads the doctrine, and its only caller anywhere is a check script. Roughly half the 179-line body (SKILL.md:33-67 and 75-128) states the same four-tag taxonomy twice, and that taxonomy already exists in longer, sharper form in `INSTRUCTIONS.md:122-166` sitting in the same working directory — under *different tag names* (`BEHAVIORAL` here vs `RULE VIOLATION` there), so the optimizer reads two overlapping vocabularies for one concept on every iteration. The body also bakes in capability-specific levers ("fix in that EXISTING tool's body", "composite atomic-WRITE tool") and a benchmark-shaped ACT-vs-REFUSE tag plus a run postmortem ("this exact mistake sank a prior run", SKILL.md:56) into what must be a capability- and benchmark-agnostic phase. `python skills/phases/diagnose/scripts/check.py` passes (`ok: true`, 5 notes).

## Findings

**F1. The deterministic clusterer is excluded from the only place the skill is actually read, and the skill's own "How to run" is a dead end there.**

`harness.py:1231-1243` copies `skills/phases/diagnose/` into `workdir/guidance/diagnose/` with `ignore=shutil.ignore_patterns("__pycache__", "scripts", "*.pyc")` (`harness.py:1240`); the native-injection path does the same (`harness.py:1295-1321`). `INSTRUCTIONS.md:85` then says `./guidance/diagnose/SKILL.md — the failure-clustering method. Use it.` So the optimizer gets 179 lines of prose doctrine and *not* `scripts/run.py`. Meanwhile SKILL.md:161-166 instructs the reader to `python scripts/run.py --run-dir ... --tag seed` and "feed `reflective_dataset` + `clusters` into the algorithm's proposal prompt" — an instruction that cannot be followed from `guidance/diagnose/`, where no `scripts/` exists. A grep for production callers finds none: the only reference to `phases/diagnose/scripts/run.py` outside this skill is `skills/algorithms/agent-optimize/scripts/check.py:57`, a test. No algorithm (`hill-climb`, `gepa`, `skillopt`) mentions diagnose at all. Net effect: every iteration of every run pays the full body cost, and the model re-derives clustering by hand in prose because the script it is nominally backed by is not there.

**Fix:** (a) drop `"scripts"` from the ignore pattern at `harness.py:1240` so `guidance/diagnose/scripts/` ships, and have the algorithm/harness *run* `run.py` and inject its JSON into the `{{FAILURES}}` block rather than asking the optimizer to re-cluster from raw traces; or (b) if the copy must stay prose-only, delete the "How to run" invocation from the copied surface and move the runnable path into a `references/` file, so the doctrine the optimizer reads never names a script it cannot execute. Either way the current state — instruct-and-withhold — is the worst of both.

**F2. The clustering procedure is hand-waved in prose, and the one concrete signature that exists both splits a single root cause into three clusters and merges three unrelated causes into one.**

*Prose side.* SKILL.md:33 ("grouped by shared root cause"), :77-79 ("Cluster **ALL** failing trajectories by shared root cause"), and `references/concepts.md:53` ("Grouping failures by a shared signature") never state how to compute a signature or what makes two failures the same cluster. The only operational guidance anywhere is a negative smell test at `concepts.md:65-66` ("'one cluster per task' means the signature is too fine"). A procedure whose entire specification is "don't make too many" is not a procedure — it is a hope, and it is re-improvised differently on every iteration.

*Script side.* `run.py:49-60` derives the signature as: lowercase → strip quoted literals → strip digits → strip punctuation → keep the first 8 tokens longer than 2 chars. That is a lexical prefix hash of the scorer's feedback string, not a root-cause key. Measured on realistic feedback:

```
'task failed the agent did not confirm the'      <- Task failed: the agent did not confirm the change...
'agent omitted required confirmation step'       <- Agent omitted required confirmation step
'missing confirmation before the write'          <- Missing confirmation before the write
```
One root cause (missing confirmation), three clusters — exactly the over-fragmentation `concepts.md:65` warns against. And in the other direction:
```
'grading failed for this trajectory because the expected'   <- ...: missing confirmation
'grading failed for this trajectory because the expected'   <- ...: wrong payment id
'grading failed for this trajectory because the expected'   <- ...: refused a valid request
```
Any scorer with a boilerplate preamble longer than 8 tokens collapses every distinct cause into one cluster. `check.py:46-50` only exercises the easy case (same sentence, different digits), so neither failure mode is caught.

**Fix:** specify the criterion, then encode the mechanical part. Concretely: a cluster is defined by a **(failure-site, violated-expectation)** pair — *where* in the trajectory it went wrong (the tool/step/field the trace names) and *what* expectation was missed — and two failures join the same cluster when both components match, independent of task-specific values and independent of the scorer's phrasing. Then split the work by determinism: (i) put in `scripts/` the deterministic half — load rollouts, strip task-specific literals, and emit *candidate* groups plus, per group, the evidence lines the model needs (drop the fixed prefix common to all feedback strings before signature extraction; that single change fixes the collapse case above); (ii) keep in SKILL.md only the judgment half — reading the trace to name the failure site and the violated expectation, and merging/splitting the candidate groups. Add the two cases above to `check.py` as regression tests: same cause under three phrasings must yield one cluster, three causes under one preamble must yield three.

**F3. Half the body states the same four-tag taxonomy twice, and the whole taxonomy already exists — longer, sharper, and with different tag names — in the file next to it.**

Internal duplication: SKILL.md:45-67 defines KNOWLEDGE / BEHAVIORAL / DECISION-PERMISSION / CAPABILITY-GAP with their fixes and blast-radius warnings; SKILL.md:84-96 defines the same four tags with the same fixes and the same warnings. `blast radius` appears 5×, `DECISION / PERMISSION` 3×, `BEHAVIORAL` 3×, `partial-credit` 2×, the no-leak rule 4× (:128, :147, :170, :173) — plus a fifth statement in `concepts.md:42-48`. That is ~89 of 179 lines spent saying one thing twice.

External duplication: `templates/project/optimizer/INSTRUCTIONS.md:122-166` already carries the full lever-by-failure-type taxonomy with worked code, `:48-64` the bounded/unbounded blast-radius discipline, `:161-162` the same "this exact mistake sank a prior run" anecdote, `:194-201` the non-overfitting rule, `:110-113` "cluster ALL failures — total, partial-credit, AND communication/omission" and leverage ranking, `:107` the parallel-subagent note. Both files are in the same workdir and both are mandatory reading. Worse, the vocabularies disagree: this skill's **BEHAVIORAL** is INSTRUCTIONS' **RULE VIOLATION**, so the optimizer must reconcile two names for one tag before it can use either. The `DECISION / PERMISSION` doctrine is *also* duplicated a third time in `skills/capabilities/system-prompt/SKILL.md:88,153`.

**Fix:** make `INSTRUCTIONS.md` the single owner of "which lever fixes which failure type" (it is the per-iteration prompt; it is where the levers are already best written), and cut SKILL.md:45-67 and :84-96 down to one sentence that points there plus the tag list by name only — adopting INSTRUCTIONS' names, not competing ones. Delete SKILL.md:129-139 entirely (steps 2-3, "ideate a DRASTIC edit" / "edit and stop"): proposing and applying edits is not diagnose's job, and `INSTRUCTIONS.md:106-120` already specifies it. State the no-leak rule once (SKILL.md:145-149 is the best-worded instance) and let `concepts.md:42-48` hold the rationale. Target: body under 70 lines, in line with its sibling phases (`evaluate` 70, `gate` 83, `finalize` 63, `baseline` 79).

**F4. Capability-specific levers, a benchmark-shaped tag, and a run postmortem are baked into a phase skill that must be agnostic to all three.**

SKILL.md:48-49 ("fix in that EXISTING tool's body (an in-code guard)") and :133-137 ("PREFER writing a new code-bearing tool over a docstring tweak … wrap a primitive … then remove the raw primitive") prescribe a **tools** capability's edit space; for a `system-prompt` or `skill-package` capability there is no tool body to guard. The `DECISION / PERMISSION (ACT vs REFUSE)` tag (:50-59, :86-90) is written around policy-governed refuse/escalate behavior — a `tau2_airline`-shaped concern promoted into the generic phase, where it consumes 10 of 179 lines and forces the same reasoning on every benchmark that has no permission model. SKILL.md:56 cites an unnamed prior run's failure ("this exact mistake sank a prior run") as doctrine; a generic phase skill should not encode one run's postmortem. SKILL.md:38 illustrates partial credit as "a write got the wrong arguments" — also tools-flavored, though as a marked example it is acceptable.

**Fix:** keep in diagnose only what is true of any capability and any benchmark: *cluster the failures, rank them by recoverable score minus regression risk, name each cluster's blast radius, and hand the tagged clusters to the capability skill for the lever.* Move the tools-specific levers to `skills/capabilities/tools/SKILL.md` (which owns that edit space) and the ACT-vs-REFUSE doctrine to `skills/capabilities/system-prompt/SKILL.md:88,153`, where two thirds of it already lives. Replace the run anecdote with the general reason (a global rule change has unbounded blast radius) — which the skill already states one clause earlier and does not need the anecdote to justify.

**F5. The reflective dataset drops the trajectory the skill promises, and never mentions the adapter method that is the format-agnostic way to get it.**

`concepts.md:24-25` documents `Generated Outputs` as "what the agent actually produced (and, where available, the trajectory: reasoning, tool calls)". `run.py:89` populates it with `ro.get("output")` only — the final output string from the rollout record. The tool calls and reasoning live in the runner's native trajectory store, whose documented, structure-agnostic accessor is `Adapter.trajectories(split, ctx)` (`core/cap_evolve/adapter.py:157-171`, "*any* structure, files in *any* format", copied verbatim to `./trajectories/`). Neither `SKILL.md` nor `run.py` mentions `trajectories()`; the skill instead says "Reads a candidate's val rollouts" (:23-24, :165), presuming the rollout record's `output` field carries enough trace to diagnose from. For an agentic runner it does not — which is precisely why the failure site (F2) cannot be derived from the current reflective dataset.

**Fix:** in SKILL.md, name `./trajectories/` (in-workdir) and `Adapter.trajectories()` (standalone) as the source of the full trace, and state that the rollout record supplies the score and feedback while the trajectory supplies the failure site — going through the adapter rather than assuming any trace schema. In `run.py`, accept an optional trajectory directory and attach the matching per-task trace path (not parsed content — the format is the adapter's business) to each reflective-dataset entry, so the model reading the entry knows which file to open.

**F6. No output template, and `argument-hint` has drifted from the script it documents.**

skill-creator requires defining output formats with an explicit template ("**Defining output formats** … ALWAYS use this exact template"). 179 lines of doctrine contain no shape for what diagnose emits — a reader learns the keys `reflective_dataset` / `clusters` / `kept_good` only from prose (:165-166) and must read `run.py:93-98` for the actual JSON. Separately, `argument-hint: "--run-dir DIR --tag CANDIDATE_ID"` (SKILL.md:5) omits both `--split {train,val}` (`run.py:105-107`) and `--cluster` (`run.py:108-109`), and SKILL.md never mentions that diagnosing train is possible — even though `run.py:34-38` and `check.py:52-75` treat that as the point of the flag. The `## Dual-mode` paragraph (:159) is byte-identical across all six phase skills (`baseline`, `evaluate`, `gate`, `finalize`, `report`, `diagnose`) except the slash-command name.

**Fix:** add a compact output template (the three keys with one example cluster entry, ~8 lines) replacing the prose at :165-166. Update `argument-hint` to `--run-dir DIR --tag CANDIDATE_ID [--split train|val] [--cluster KEY]` and add one line on when to diagnose train vs val. Move the Dual-mode paragraph into one shared location (e.g. `skills/phases/README.md` or the `using-cap-evolve` router) that the six phase bodies point to in one clause.

**F7. Frontmatter is compliant; two nits.**

`name` + `description` are present and well-formed (SKILL.md:2-3); the description states both what it does and when to use it ("Use between evaluation and proposing edits"), avoids CRITICAL/ALWAYS/MUST, and carries triggering keywords (traces, reflective dataset, clusters). It opens imperatively ("Extract the learning signal…") rather than in third person — consistent with the other phase skills, so fix repo-wide or not at all. `sources: [gepa, skillgrad, trace2skill, evo]` (:9) is parsed by no code in `core/` and duplicates provenance already in `concepts.md:76-84`; `needs`/`provides`/`component`/`name` are duplicated between the frontmatter and `meta.yaml:1-8` (currently in sync). `references/concepts.md` (84 lines) is one level deep with an explicit pointer (:176-179) — correct per skill-creator.

**check.py result:** passes — `{"skill": "diagnose", "ok": true, "problems": []}` with 5 notes. Note what it does not cover: it never tests that failures with the *same* root cause under *different* phrasings cluster together (only same-sentence-different-digits, `check.py:25-28,46-50`), which is why F2's fragmentation is invisible to it.

## Acceptance criteria

1. SKILL.md body is under ~70 lines, with the taxonomy stated exactly once and no duplicate of material already in `templates/project/optimizer/INSTRUCTIONS.md`; tag names match INSTRUCTIONS' names.
2. The clustering criterion is stated concretely — what a signature is derived from, and the explicit test for whether two failures belong to the same cluster — not "by shared root cause".
3. The deterministic half of clustering runs from `scripts/` and is reachable in production: either `harness.py:1240` no longer excludes `scripts/`, or SKILL.md's copied surface no longer instructs the reader to run a script that isn't shipped there. No prose instruction names an unavailable path.
4. `check.py` gains two cases: one root cause under ≥3 different phrasings yields one cluster; ≥3 distinct causes sharing a long feedback preamble yield ≥3 clusters. Both pass.
5. No tools-capability lever, no ACT-vs-REFUSE policy doctrine, and no run-specific postmortem remain in the phase body; each has one owner in the relevant capability skill.
6. SKILL.md names `./trajectories/` / `Adapter.trajectories()` as the trace source and assumes no trace schema; each reflective-dataset entry can be traced to its full trajectory.
7. An explicit output template for the emitted JSON is present, and `argument-hint` lists `--split` and `--cluster`.
8. `python skills/phases/diagnose/scripts/check.py` still exits 0.

## Out of scope

- Redesigning `INSTRUCTIONS.md` itself, beyond making it the single owner of the lever taxonomy.
- The shared `## Dual-mode` / `## Inputs / outputs` / `## What good vs bad looks like` boilerplate in the other five phase skills (F6 names the shared-location fix; executing it across all six belongs in one cross-skill issue).
- Changing the run-dir rollout record schema, the gate, or any algorithm's loop.
- Deleting `core/build/lib/cap_evolve/` (stale build artifact, untouched here).
